repository/src/main/java/org/alfresco/repo/jscript/AlfrescoContextFactory.java

```diff
@@ -137,9 +128,40 @@ protected Object doTopCall(Callable callable, Context cx, Scriptable scope, Scri
     {
         AlfrescoScriptContext acx = (AlfrescoScriptContext) cx;
         acx.setStartTime(System.currentTimeMillis());
+        setMaxStackDepth(acx, acx.isLimitsEnabled());
         return super.doTopCall(callable, cx, scope, thisObj, args);
     }
 
+    private void setMaxStackDepth(AlfrescoScriptContext acx, boolean enable)
+    {
+        if (enable)
+        {
+            // Max stack depth
+            if (maxStackDepth > 0 && maxStackDepth != acx.getMaximumInterpreterStackDepth())
+            {
+                LOGGER.debug("Max call stack depth limit will be enabled with value: " + maxStackDepth);
+
+                if (optimizationLevel != INTERPRETIVE_MODE)
+                {
+                    LOGGER.debug("Changing optimization level from " + optimizationLevel + " to " + INTERPRETIVE_MODE);
+                }
+                // stack depth can only be set in interpretive mode
+                acx.setOptimizationLevel(INTERPRETIVE_MODE);
+                acx.setMaximumInterpreterStackDepth(maxStackDepth);
+            }
+        }
+        else
+        {
+            if (acx.getMaximumInterpreterStackDepth() != Integer.MAX_VALUE)
+            {
+                LOGGER.debug("Max call stack depth limit will be set to default value: " + Integer.MAX_VALUE);
+                acx.setOptimizationLevel(INTERPRETIVE_MODE);
```

Comment:
setting OptimizationLevel twice, here to INTERPRETIVE_MODE and two lines below to optimizationLevel

Reply:
@nbarithel the reason behind setting the optimization level twice is to prevent an exception to be thrown when changing the max stack depth to the default value. Please, see [here](https://github.com/mozilla/rhino/blob/0131b5c7e4b3924241e0a6a7eb2d52ed7e888f89/src/org/mozilla/javascript/Context.java#L1986).

Reply:
Indeed, thanks for the link, it wasn't obvious from reading the code! Although, it seems that the maximum stack depth setting will still not work if the optimization level is set to something different from -1 afterwards, according to the [javadoc](https://github.com/mozilla/rhino/blob/0131b5c7e4b3924241e0a6a7eb2d52ed7e888f89/src/org/mozilla/javascript/Context.java#L1974).

Reply:
Ok, now I get it, you just wanted to reset the default value of the maximumInterpreterStackDepth 👍